Tutorial | Custom Shape Tool in C++ and Python (#1119)
o3de/o3de.org: content/docs/learning-guide/tutorials/extend-the-editor/shape-example-py.md

---
linkTitle: Create a Custom Tool Gem in Python
title: Create a Custom Tool Gem in Python to Extend Open 3D Engine Editor
description: Learn how to extend the Open 3D Engine (O3DE) Editor by creating a custom tool Gem in Python.
weight: 300
toc: true
---

> In this tutorial, you create a custom tool Gem written in Python, called **Shape Example**, that extends the **Open 3D Engine (O3DE) Editor**. The Shape Example tool allows you to create entities with a Shape component and configure their component properties. You'll learn how to use the `PythonToolGem` template and practice Python development with [Qt](https://wiki.qt.io/Main), O3DE Tools UI API, and other O3DE APIs.

Reviewer: Many of the same comments in the cpp tutorial apply to the Python tutorial, too.

This tutorial shows you how to create the **PyShapeExample** Gem, which is a sample Gem in [`o3de/sample-code-gems` repository](https://github.com/o3de/sample-code-gems/tree/main/py_gems/PyShapeExample). You can reference the PyShapeExample Gem sample as you follow along this tutorial.

By the end of this tutorial, you'll be able to extend the Editor by creating your own custom tools written in C++.

The following image is a preview of the Shape Exmple tool that you create in this tutorial.

{{< image-width "/images/learning-guide/tutorials/extend-the-editor/shape-example/python-shape-example-demo.png" "1080" "An image of the Shape Example tool and some entities created by it." >}}

## Prerequisites

Before you start the tutorial, ensure that you have the following:

> - Set up an O3DE development environment. You can build O3DE as a source engine or a pre-built SDK engine. For more information, refer to [Setting up O3DE from GitHub](/docs/welcome-guide/setup/setup-from-github.md).

Reviewer: ```suggestion
- Set up an O3DE development environment. For instructions, refer to [Set up Open 3D Engine](/docs/welcome-guide/setup/).
```

## Create a Gem with the `PythonToolGem` template

To start, create a Gem by using the `PythonToolGem` template. The `PythonToolGem` template provides a basic Python framework to create a dialog window that extends the Editor.

This tutorial uses the following directory names and locations in the examples.

> - **O3DE engine directory**: `C:/o3de`  

Reviewer: Same comment from the C++ tutorial about consistent use of back slashes applies to this Python tutorial.

    ```

   Depending on the Gem path, this command automatically registers the Gem to one of the manifest files: `.o3de/o3de_manifest.json`, `<engine>/engine.json`, and `<project>/project.json`.

2. (Optional) Register the Gem to your project. This step is optional because when you create a Gem using the O3DE CLI in the previous step, it automatically registers the Gem.

    ```cmd
    scripts/o3de register -gp C:/o3de-gems/MyPyShapeExample -espp <project-path>
    ```

3. Add the Gem in your project by using the `o3de` script.

    ```cmd
    scripts/o3de enable-gem -gn MyPyShapeExample -pp <project-path>
    ```

    Or, enable the Gem using the Project Manager (refer to [Adding and Removing Gems in a Project](/docs/user-guide/project-config/add-remove-gems.md)).

4. Build the project by using the `o3de` script (refer to [Build a project](/docs/user-guide/build/configure-and-build/#build-a-project)). Or, use the Project Manager (refer to the **Build** action in the [Project Manager](/docs/user-guide/project-config/project-manager/)) page.

5. Open Editor for your project.

> 6. Open the tool by selecting **Tools > Examples > MyPythonShapeExample** from the file menu. (See A in the following image.)

Reviewer: ```suggestion
6. Open the tool by selecting **Tools > Examples > MyPyShapeExample** from the file menu. (See A in the following image.)
```

Example of `Code/Source` directory:

{{< image-width "/images/learning-guide/tutorials/extend-the-editor/shape-example/python-tool-gem-template-directory.png" "720" "Editor with a tool created using the PythonToolGem template." >}}

Example of `Editor/Scripts` directory:

{{< image-width "/images/learning-guide/tutorials/extend-the-editor/shape-example/python-tool-gem-template-directory-2.png" "620" "Editor with a tool created using the PythonToolGem template." >}}

### Modules and system components

All Gems have modules and system component classes written in C++ that connect the tool to O3DE and allow it to communicate with other systems. The `PythonToolGem` template already contains all of the code needed to run the tool in the Editor. You can find these C++ files in `Code/Source`.

> For more information on Gem modules and system components, refer to the [Overview of the Open 3D Engine Gem Module system](/docs/user-guide/programming/gems/) page.

Reviewer: ```suggestion
For more information on Gem modules and system components, refer to the [Overview of the Open 3D Engine Gem Module system](/docs/user-guide/programming/gems/overview/) page.
```

An input field takes text input from the user. With Qt, you can create an input field by using the `QLineEdit` object. In this example, the input field is used to name the generated entity. You will define this behavior later.

1. Create an input field by instantiating `QLineEdit`. In this example, name it `name_input`.

2. Set the placeholder text in `name_input` by calling `setPlaceholderText(...)`.

3. Enable a button to clear the text using `setClearButtonEnabled(True)`.

```py
        self.name_input = QLineEdit(self)
        self.name_input.setPlaceholderText("Optional")
        self.name_input.setClearButtonEnabled(True)
```

### Create a checkbox

A checkbox is an option button that users can enable or disable to trigger a user-defined behavior. With Qt, you can create an checkbox by using the `QCheckBox` object. In this example, the checkbox controls whether or not to append a suffix to the entity's name. At runtime, it will start disabled, and enable when the user enters the entity's name to the input field. You will define this behavior later.

1. Create a checkbox by instantiating `QCheckBox`. In this example, name it `add_shape_name_suffix`.

> 2. Set the checkbox to the start in the disabled state by using `setDisabled(True)`.

Reviewer: ```suggestion
2. Set the checkbox to start in the disabled state by using `setDisabled(True)`.
```

1. Add an image file to the `Code/Source` directory to use as your icon. In this example, the icon is named `toolbar.svg`. We recommend that your icon adheres to the guidelines in [UI development best practices](https://o3de.org/docs/tools-ui/uidev-component-development-guidelines/#ui-development-best-practices).

2. Add your icon to PyShapeExample Gem's resources by updating `Code/Source/ShapeExample.qrc` with your new icon's file name.

    ```xml
    <!DOCTYPE RCC><RCC version="1.0">
        <qresource prefix="/PyShapeExample">
            <file alias="toolbar_icon.svg">toolbar_icon.svg</file>
        </qresource>
    </RCC>
    ```

3. Register PyShapeExample Gem's resources to Qt Resource System by adding the following code in `Code/Source/EditorModule.cpp`. 

    - Define the function `InitShapeExampleResource()` and call `Q_INIT_RESOURCE(...)` to register the Qt resrouces listed in `ShapeExample.qrc`.

        ```cpp
        void InitShapeExampleResources()
        {

>             Q_INIT_RESOURCE(ShapeExample);

Reviewer: PyShapeExample?
```suggestion
            Q_INIT_RESOURCE(PyShapeExample);
```

1. Open Editor for your project.

2. Open the Python Scripts panel by selecting **Tools > Other > Python Scripts** from the file menu.

3. Open `pyshapeexample_dialog` in the Python Scripts panel.

Congratulations! You created a custom tool Gem that's written in Python, built it, and loaded it in the Editor. Your Shape Example tool should look something like this:

{{< image-width "/images/learning-guide/tutorials/extend-the-editor/shape-example/python-shape-example-ui.png" "500" "An image of the Shape Example tool." >}}

## Download the PyShapeExample Gem sample

This tutorial is based off of the PyShapeExample Gem in [`o3de/sample-code-gems` repository](https://github.com/o3de/sample-code-gems/tree/main/py_gems/PyShapeExample). You can download this sample and load it in the Editor.

1. Download or clone the repository. The PyShapeExample Gem is located in `sample-code-gems/py_gems/PyShapeExample`.

    ```cmd
    git clone https://github.com/o3de/sample-code-gems.git
    ```

2. Before you can open the Shape Example tool, do the following:

   - Register your Gem.

   - Enable it in your project.

   - Rebuild your project.

   - Open the tool in the Editor.

>     These steps are explained earlier in this tutorial. Refer to [Create a tool with the `PythonToolGem` template](#create-a-tool-with-the-pythontoolgem-template).

Reviewer: ```suggestion
    These steps are explained earlier in this tutorial. Refer to [Create a Gem with the `PythonToolGem` template](#create-a-gem-with-the-pythontoolgem-template).
```
